### PR TITLE
Finish Backend MFA Setup

### PR DESCRIPTION
**Pull request by emmastephenson**
_Note: I would appreciate reviews from all requested reviewers!_
This is a massive PR, and you may be wondering why I requested your review! I'm going to list the specific sections I'd like folks to review below, because you're all on here for a reason. That said, please do review the entire PR if you have time, and don't feel limited to the things I've called out below - the more eyes on this, the better!
Ben - the new Gradle dependency (`samstevens:totp`); also would appreciate any thoughts you have on how to integrate `LiveOktaAuthenticationTest` into our flow without running it too often & how to get the Okta `token` into our test env
Brett - `LiveOktaAuthenticationTest`, as we chatted about integration testing Okta a few weeks ago
Nick - checking the endpoints to ensure that they're compatible with what the frontend is expecting
Tim - general sanity check that this is how we want the account creation flow to behave

Thank you all!!

## Related Issue or Background Info

#1559 
#1548 
#1547 
#1552 

## Changes Proposed

- Adds endpoints that ping Okta for enrolling authentication apps, verifying a user-input passcode, and resending a passcode. 
- Updates UserAccountCreationControllerTest to utilize helper methods for creating requests.
- Updates DemoOktaAuthentication with corresponding enrollment methods.
- Adds LiveOktaAuthenticationTest, which actually creates a user in Okta and goes through the account enrollment flow.

## Additional Information

- [DO NOT SUBMIT] This PR shouldn’t be submitted as-is, because we need to add the Okta token to our test runner (this is the only way to get LiveOktaAuthenticationTest to pass - to test it locally, I’ve been directly adding the token to `application-default.yaml`, but this is a bad long-term solution)
- I added LiveOktaAuthenticationTest because I didn’t quite trust the Okta SDK to perform as expected, and because there are a couple cases where I have to parse a JSON response from Okta, so testing seemed like a good idea :) 
- This introduces a dependency on `samstevens:totp`, a library used only in LiveOktaAuthenticationTest for verifying a passcode (Okta doesn’t tell us what passcode it sent to the user in any of the API responses, so the only way to test that was to re-create the passcode using the shared secret they do return)

## Testing
`LiveOktaAuthenticationTest`
On test initialization, creates a user and an instance of `LiveOktaAuthentication`. (Creating a user here is the equivalent of clicking "add user" in-app.) Uses test order to enforce that everything runs in the correct order, otherwise tests will be very flaky (i.e., the test for adding an MFA option will fail if the user hasn't been initialized yet.) Runs through the account setup flow, first setting password, then recovery questions, then enrolling the user in MFA options. (The MFA enrollment is deleted at the end of each test, so a user is only enrolled in one MFA option at a time.)
To verify that `LiveOktaAuthentication` is WAI, the test makes direct calls to the Okta API, largely to get the user or get their factors. These API calls are also how a user is created and deleted in test setup/teardown.

### UI
- ~~[ ] Any changes to the UI/UX are approved by design~~
- ~~[ ] Any new or updated content (e.g. error messages) are approved by design~~

### Testing
- [x] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- ~~[ ] Database changes are submitted as a separate PR~~
  - ~~[ ] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user~~
  - ~~[ ] Any changes to tables that have custom no-PHI views are accompanied by changes to those views~~
- ~~[ ] GraphQL schema changes are backward compatible with older version of the front-end~~

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

## Cloud
- [ ] DevOps team has been notified if PR requires ops support
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test` environment for verification

### CODE REVIEW

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+              _auth.resendActivationPasscode(_userId, "unenrolledFactorId");
+            });
+    assertThat(exception).hasMessage("An exception was thrown while fetching the user's factor.");
+  }
+
+  /**
+   * Helper method to create a user.
+   * https://developer.okta.com/docs/reference/api/users/#request-example
+   */
+  private JSONObject createUser() {
```
Comment (bmaden-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java:360]: Since this test class already has the Okta token and org url, I think it could just instantiate an Okta `Client` and use it for what `createUser`, `activateUser`, `deactivateUser`, etc are doing.  Achieves the same thing, but might make these helpers  unnecessary.

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+ * creation flow. The user is not associated with an organization, and is deleted at the end of the
+ * test. Running this test does add to our Okta quota, so plase run sparingly.
+ */
+@TestInstance(Lifecycle.PER_CLASS)
+@TestMethodOrder(OrderAnnotation.class)
+class LiveOktaAuthenticationTest extends BaseFullStackTest {
+
+  private static final String PHONE_NUMBER = "999-999-9999";
+  private static final String FORMATTED_PHONE_NUMBER = "+19999999999";
+  private static final String EMAIL = "test@example.com";
```
Comment (bmaden-usds) [inline backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java:46]: If we do come up with a way to run these automatically, the fixed email address will prevent it from being run successfully from multiple clients concurrently.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/UserAccountCreationController.java [hunk 3]
```diff
@@ -134,8 +135,13 @@ public void enrollEmailMfa(@RequestBody EnrollMfaRequest requestBody, HttpServle
    * @throws OktaAuthenticationFailureException if Okta cannot enroll the user in MFA.
    */
   @GetMapping("/authenticator-qr")
-  public void getAuthQrCode(@RequestBody EnrollMfaRequest requestBody, HttpServletRequest request) {
-    // WIP: doesn't interact with Okta yet.
+  public String getAuthQrCode(
```
Comment (bmaden-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/UserAccountCreationController.java:138]: Should we return a json body with a property for the qr code link?  If it's just a text response with the string then it'll be more difficult to change if we ever need to (for example, adding something to the response would become a breaking change).

**Review by bmaden-usds — CHANGES_REQUESTED** [review]
Nice!  You've done a lot of great work!

For now, I think the LiveOktaAuth tests will have to be disabled.  Conditionally would be preferable.

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+/**
+ * WARNING: THIS TEST CREATES A REAL USER IN OKTA!
+ *
+ * <p>This tests LiveOktaAuthentication by creating a user and sending them through the account
+ * creation flow. The user is not associated with an organization, and is deleted at the end of the
+ * test. Running this test does add to our Okta quota, so plase run sparingly.
+ */
+@TestInstance(Lifecycle.PER_CLASS)
+@TestMethodOrder(OrderAnnotation.class)
+class LiveOktaAuthenticationTest extends BaseFullStackTest {
```
[the remark this answers is not in the document]
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java:42]: I didn't know about that annotation, thank you! That's a great idea. I'd prefer to enable it only if `LiveOktaAuthentication` has been changed, but it doesn't look like that's feasible with Spring.
I think the next-best option would be to only run this during deploys to `stg`, but I'm not familiar enough with our test setup to know if that's possible using the `@EnabledIf` annotation either :) If it's not, then my inclination would be to just `@Disable` the test entirely and throw some warning comments in `LiveOktaAuthentication` that if you're changing that code, you also need to run `LiveOktaAuthenticationTest`.

File: backend/src/test/resources/application-default.yaml [hunk 5]
```diff
@@ -30,6 +30,10 @@ logging:
     # Other possibilities:
     # com.okta: DEBUG
     # org.springframework.security: DEBUG
+okta:
+  client:
+    org-url: https://hhs-prime.okta.com
+    token: ${OKTA_API_KEY:MISSING}
```
[the remark this answers is not in the document]
Comment (emmastephenson) [reply backend/src/test/resources/application-default.yaml:36]: Oh very good point, I think I put this here at some point when I was attempting to get the test to build. Removed, and I'll add some validation in the test initialization to make sure the test doesn't run if the token is empty.

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+  private LiveOktaAuthentication _auth;
+  private RestTemplate _restTemplate;
+  private String _userId;
+  private String _activationToken;
+  private String _apiUrl;
+  private HttpEntity<String> _entityWithHeaders;
+
+  @BeforeAll
+  void initializeUser() {
+    _auth = new LiveOktaAuthentication(_orgUrl, _token);
```
[the remark this answers is not in the document]
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java:70]: Good point, added an IllegalArgumentException check.

File: backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/UserAccountCreationController.java (hunk 3 above)
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/apiuser/UserAccountCreationController.java:138]: Sounds good, changed to a JSONObject. (In `UserAccountCreationController` I'm still just sending the QR code link to the frontend though.)

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java (hunk 1 above)
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java]: Done - but a word of warning, this has made the diff look much worse than it actually is 😄  The test assertions haven't actually changed, I'm just asserting on Java objects instead of JSON now.

File: backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java (hunk 2 above)
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/idp/authentication/LiveOktaAuthenticationTest.java:57]: Hmmm, that's a good point. I wonder if it'd be worth doing something like 

```
String now = Instant.now().toString();
String EMAIL = "test+" + now + "@example.com"
```

which would make some ugly email addresses, but would hopefully be unique enough that it could run from multiple clients.

**Review by nickclyde — APPROVED** [review]
I approve this since it now works in test! Also thanks for adding the TOTP test!!

**Comment by nullflux** [comment]
I gather that this PR is obsolete since #1817 and can be closed?

**Comment by emmastephenson** [comment]
@nullflux yep, thanks for catching!
